Implement HexagonMeshTrimmer

miaoyinb opened the pull request:
closes #21853

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->

## Design
<!--A concise description (design) of the enhancement.-->

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

miaoyinb committed:
Implement HexagonMeshTrimmer

closes #21853

clang-format fix ref #21853

CIVET compilation warning fix ref #21853

clang-format fix 2 ref #21853

miaoyinb commented:
@eshemon

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1144537/) on da24c29 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21863/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21863/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8606aec106f76cb69b6d688124abd738582261ac`

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 937327ae580cea63ce6ff0e4cc5b9d64160f48e0
+Subproject commit 8276e3d03117ff4836da29d16d2b85c26866c9e6
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1144651/) on accb59c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21863/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21863/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8606aec106f76cb69b6d688124abd738582261ac`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1144708/) on 1cf6698 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21863/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1144741/) on 1cf6698 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>8606ae</th>
      <th colspan=4>#21863 1cf669</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.11%</td>
      <td>84.11%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>78645</td>
      <td>78644</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14857</td>
      <td>14858</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21863/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21863/coverage/framework)


## Modules coverage

### Reactor

<table>
  <thead>
    <tr>
      <th></th>
      <th>8606ae</th>
      <th colspan=4>#21863 1cf669</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.61%</td>
      <td>95.53%</td>
      <td>-0.07%</td>
      <td>94.65%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3503</td>
      <td>3785</td>
      <td>+282</td>
      <td>283</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>161</td>
      <td>177</td>
      <td>+16</td>
      <td>16</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21863/coverage_diff/reactor)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21863/coverage/reactor)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21863/coverage/xfem)

</details>



This comment will be updated on new commits.

miaoyinb commented:
@GiudGiud @loganharbour 

This is to add trimming feature to assembly and core meshes. Thank you in advance!

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/patterned_trimmed.i:
```diff
@@ -0,0 +1,78 @@
+  [hex_1]
+    type = PolygonConcentricCircleMeshGenerator
+    num_sides = 6
+    num_sectors_per_side = '4 4 4 4 4 4'
```
```suggestion
    num_sectors_per_side = '1 1 1 1 1 1'
```
let s cut the solution file size

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/assembly_trim.i:
```diff
@@ -0,0 +1,55 @@
+  [hex_1]
+    type = PolygonConcentricCircleMeshGenerator
+    num_sides = 6
+    num_sectors_per_side = '4 4 4 4 4 4'
+    background_intervals = 2
+    ring_radii = 4.0
+    ring_intervals = 2
```
```suggestion
    num_sectors_per_side = '1 1 1 1 1 1'
    background_intervals = 1
    ring_radii = 4.0
    ring_intervals = 1
```

miaoyinb replied:
2 is the minimum value for `num_sectors_per_side`

miaoyinb committed:
reduce glod files size ref. #21853

gold file size reduction, test improvement, typo fixes ref. #21853

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/assembly_trim.i:
```diff
@@ -0,0 +1,55 @@
+[Mesh]
+  final_generator = trim
```
it d be better to set the active/inactive.

rn you run a bunch of extra generators and discard the results

GiudGiud commented on modules/reactor/src/meshgenerators/PatternedHexMeshGenerator.C:
```diff
@@ -594,14 +603,24 @@ PatternedHexMeshGenerator::generate()
     }
   }
 
+  auto side_list = out_mesh->get_boundary_info().build_side_list();
```
why is this removing boundary id, along with the previous change

should we have a routine to do this if the need is shared??

miaoyinb replied:
We will need to develop a better/systematic boundary management system within the `Reactor` module. (maybe FY23)

GiudGiud commented on modules/reactor/src/meshgenerators/PatternedHexPeripheralModifier.C:
```diff
@@ -67,6 +67,9 @@ PatternedHexPeripheralModifier::PatternedHexPeripheralModifier(const InputParame
     _is_control_drum_meta(declareMeshProperty<bool>("is_control_drum_meta", false)),
+    _hexagon_peripheral_trimmability(
+        declareMeshProperty<bool>("hexagon_peripheral_trimmability", false)),
+    _hexagon_center_trimmability(declareMeshProperty<bool>("hexagon_center_trimmability", false)),
```
they look center-trimmable to me?
what's the hold up?

miaoyinb replied:
This generator is to modify the peripheral region of an assembly to help stitch it with other assemblies using HexPatternMG.

We do not need to stitch (also cannot) a center-trimmed assembly with other assemblies. So, No point to do this.

GiudGiud commented on modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:
```diff
@@ -126,13 +126,19 @@ PeripheralRingMeshGenerator::PeripheralRingMeshGenerator(const InputParameters &
+    _input(getMeshByName(_input_name)),
+    _hexagon_peripheral_trimmability(
+        declareMeshProperty<bool>("hexagon_peripheral_trimmability", false)),
+    _hexagon_center_trimmability(declareMeshProperty<bool>("hexagon_center_trimmability", false))
```
they look center trimmable to me?

miaoyinb replied:
There is no need to trim a single pin.

GiudGiud commented on modules/reactor/include/meshgenerators/HexagonMeshTrimmer.h:
```diff
@@ -0,0 +1,158 @@
+  const BoundaryName _peripheral_trimming_section_boundary;
+  /// Sector number to remain
+  const unsigned int _center_trim_sector_number;
+  /// Starting Sector to remain after
```
```suggestion
  /// Index of the sector to start trimming from (counter-clockwise direction)
```

GiudGiud commented on modules/reactor/include/meshgenerators/HexagonMeshTrimmer.h:
```diff
@@ -0,0 +1,158 @@
+  const BoundaryName _external_boundary_name;
+  /// SubdomainName suffix used to rename the converted triangular elements
+  const SubdomainName _tri_elem_subdomain_name_suffix;
+  /// Customized id shift to define subdomain ids of the converted triangular elements
+  const subdomain_id_type _tri_elem_subdomain_shift;
```
having a shift seems less convenient than:
- specify id directly
- if unspecified, use next available using that new API that I think Shikhar added

miaoyinb replied:
The users do not necessarily know whether or how many tri-element subdomains will be generated due to trimming. So shift would be a more convenient approach. If this is not specified, max_subdomain_id will be used as the shift to ensure uniqueness.

GiudGiud commented on modules/reactor/include/meshgenerators/HexagonMeshTrimmer.h:
```diff
@@ -0,0 +1,158 @@
+protected:
+  /// Input mesh to be modified
+  const MeshGeneratorName _input_name;
+  /// Peripheral regions that need to be trimmed
```
```suggestion
  /// Index of the peripheral regions to be trimmed (see moosedocs for indexing scheme)
```

GiudGiud reviewed (changes requested):
all the gold files have 15 unused nodesets and sidesets. This doesnt look so good when opening exodus files.
We should clear them at some point in the generation.

miaoyinb commented on modules/reactor/doc/content/source/meshgenerators/HexagonMeshTrimmer.md:
```diff
@@ -0,0 +1,44 @@
+
+### Peripheral Trimming
+
+ Peripheral trimming trims off peripheral region(s) of one or multiple sides of the input hexagonal assembly mesh. To be specific, for each side, half of the unit pin meshes are trimmed off, as shown by the purple lines in [hexagon_trimmer]. Each side of the hexagonal assembly is assigned an index as illustrated in [hexagon_trimmer]. Users can use [!param](/Mesh/HexagonMeshTrimmer/trim_peripheral ...
```
This for applying mirror boundary conditions

GiudGiud replied:
sounds good for assemblies yeah

miaoyinb commented on modules/reactor/src/meshgenerators/PatternedHexMeshGenerator.C:
```diff
@@ -154,10 +155,14 @@ PatternedHexMeshGenerator::PatternedHexMeshGenerator(const InputParameters & par
         "control_drums_azimuthal_meta", std::vector<std::vector<Real>>())),
     _position_file_name(declareMeshProperty<std::string>("position_file_name",
                                                          getParam<std::string>("position_file"))),
+    _hexagon_peripheral_trimmability(
+        declareMeshProperty<bool>("hexagon_peripheral_trimmability", !_generate_core_metadata)),
```
Peripheral trimming is to enable meshing of assemblies with pins on their boundaries. Thus, a reactor core should never be trimmed peripherally.

GiudGiud reviewed (changes requested):
one exodus file is present / modified in two commits
[patterned_trimmed_in.e](https://github.com/idaholab/moose/pull/21863/commits/78af3cd9e6ea9d9fd0b0976096b5954e24f23170#diff-22b41965f9df4054e346a8f303e51463d7ad93a4241e02152362c92039647d7d)
also squash the clang formats with their respective commits

is there documentation that lists for the reactor modules:
- mesh generators that build up the mesh
- operations to modify the meshes and for what reason
this is to avoid users sifting through the source index to find what operations are actually available

eshemon commented:
> one exodus file is present / modified in two commits [patterned_trimmed_in.e](https://github.com/idaholab/moose/pull/21863/commits/78af3cd9e6ea9d9fd0b0976096b5954e24f23170#diff-22b41965f9df4054e346a8f303e51463d7ad93a4241e02152362c92039647d7d) also squash the clang formats with their respective commits
> 
> is there documentation that lists for the reactor modules:
> 
> * mesh generators that build up the mesh
> * operations to modify the meshes and for what reason
>   this is to avoid users sifting through the source index to find what operations are actually available

Not yet. 
We can add a simple list easily.
Next year we will be building a comprehensive tutorial.

miaoyinb commented:
> one exodus file is present / modified in two commits [patterned_trimmed_in.e](https://github.com/idaholab/moose/pull/21863/commits/78af3cd9e6ea9d9fd0b0976096b5954e24f23170#diff-22b41965f9df4054e346a8f303e51463d7ad93a4241e02152362c92039647d7d) also squash the clang formats with their respective commits
> 
> is there documentation that lists for the reactor modules:
> 
> * mesh generators that build up the mesh
> * operations to modify the meshes and for what reason
>   this is to avoid users sifting through the source index to find what operations are actually available

I think it is due to the changes in mesh density to reduce gold file size. Will squash commits.

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,566 @@
+      "trim_peripheral_region",
+      std::vector<unsigned short>(6, 0),
+      "trim_peripheral_region<=1",
+      "Whether the peripheral region on each of the six sidesto will be trimmed in an assembly "
```
```suggestion
      "Whether the peripheral region on each of the six sides will be trimmed in an assembly "
```

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,566 @@
+  params.addRequiredParam<MeshGeneratorName>("input", "The input mesh that needs to be trimmed.");
+  params.addRangeCheckedParam<std::vector<unsigned short>>(
+      "trim_peripheral_region",
+      std::vector<unsigned short>(6, 0),
```
should be bools

miaoyinb replied:
It is easier to check a short vector than a bool vector for error messages. Also vector<bool> is not a supported input parameter time.

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,566 @@
+      std::vector<unsigned short>(6, 0),
+      "trim_peripheral_region<=1",
+      "Whether the peripheral region on each of the six sidesto will be trimmed in an assembly "
+      "mesh."); // need a test
```
```suggestion
      "mesh. See documentation for numbering convention.");
```

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,566 @@
+      "Whether the peripheral region on each of the six sidesto will be trimmed in an assembly "
+      "mesh."); // need a test
+  params.addParam<BoundaryName>("peripheral_trimming_section_boundary",
+                                "Boudary formed by peripheral trimming.");
```
```suggestion
                                "Boundary formed by peripheral trimming.");
```

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,566 @@
+                                            "center_trim_ending_index<12",
+                                            "Index of the ending center trimming position.");
+  params.addParam<BoundaryName>("center_trimming_section_boundary",
+                                "Boudary formed by center trimming.");
```
```suggestion
                                "Boundary formed by center trimming.");
```

GiudGiud replied:
there's spellchecker plugins in VSCode if you're using that btw

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,566 @@
+      "Customized id shift to define subdomain ids of the converted triangular elements.");
+
+  params.addClassDescription(
+      "This HexagonMeshTrimmer object perform peripheral and/or across-center trimming for "
```
```suggestion
      "This HexagonMeshTrimmer object performs peripheral and/or across-center (0, 0, 0) trimming for "
```

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,566 @@
+
+  params.addClassDescription(
+      "This HexagonMeshTrimmer object perform peripheral and/or across-center trimming for "
+      "assembly or core mesh generated by PatternedHexMG.");
```
```suggestion
      "assembly or core 2D meshes generated by PatternedHexMG.");
```

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,566 @@
+}
+
+void
+HexagonMeshTrimmer::quasiTriElementsFixer(ReplicatedMesh & mesh,
```
needs more comments for maintainability

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,566 @@
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+  const subdomain_id_type max_subdomain_id(*subdomain_ids_set.rbegin());
+  const subdomain_id_type tri_subdomain_id_shift =
+      _tri_elem_subdomain_shift == Moose::INVALID_BLOCK_ID ? max_subdomain_id
```
add an assert that 2x max_subdomain_id is not too big

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,566 @@
+  }
+  unsigned int bad_elem_ext = 0;
+  std::set<subdomain_id_type> new_subdomain_ids;
+  for (unsigned i = 0; i < bad_elem_counter; i++)
```
i is only used to access the element right? so you could use a range-based iterator?

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,566 @@
+                 "subdomain name: " +
+                 mesh.subdomain_name(nid) + ".");
+  }
+  mesh.prepare_for_use();
```
this can be expensive.  let's move it up in generate() so we dont have to search for it

miaoyinb committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

miaoyinb committed:
Add comments and improve coding ref #21853

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,578 @@
+  // move nodes
+  for (unsigned int i = 0; i < node_list.size(); i++)
+  {
+    if (node_list_flag[i])
```
```suggestion
    // Only one node in trimmed region
    if (node_list_flag[i])
```

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,578 @@
+  {
+    if (node_list_flag[i])
+      *(mesh.node_ptr(node_list[i])) = node_list_point[i];
+    else
```
```suggestion
    // Two nodes in trimmed region, only one is moved
    else
```

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,578 @@
+                                (p1(0) - p0(0)) / (p0(1) - p1(1)),
+                                -(p0(0) + (p1(0) - p0(0)) / (p0(1) - p1(1)) * p0(1)));
+      }
+    }
```
what about the case when there are 3 nodes of a QUAD in the trimmed region. This is the degenerate quad which turns into a TRI right?

document the node movement involved (all 3 right?) here please

miaoyinb replied:
Here we are working with nodes of a certain `side`. So the maximum number is 2.

miaoyinb committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1162992/) on 363810d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21863/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21863/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0366d364f69d8e894e7c0a4f22cfadd090a1187c`

miaoyinb commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,578 @@
+
+  for (unsigned int i = 0; i < bdry_pars.size(); i++)
+    if (trim_peripheral_region[i])
+      lineRemover(mesh,
+                  bdry_pars[i],
+                  block_id_to_remove,
+                  subdomain_ids_set,
+                  peripheral_trimming_section_boundary_id,
+                  external_boundary_id);
```
It is possible if the trimming lines intersect within the original mesh.

miaoyinb committed:
more comments and code improvement ref #21853

miaoyinb commented:
There seems to be a bug about center trimming index. I will fix it.

miaoyinb committed:
fix a center index parameter bug ref #21853

miaoyinb commented:
> There seems to be a bug about center trimming index. I will fix it.

fixed

miaoyinb commented:
@GiudGiud Let me know if any additional revision is needed. Thanks.

GiudGiud commented on modules/reactor/include/meshgenerators/HexagonMeshTrimmer.h:
```diff
@@ -0,0 +1,162 @@
+  /**
+   * Calculates the internal angles of a given 2D element
+   * @param elem the element that needs to be investigated
+   * @return sizes of all the internal angles
```
```suggestion
   * @return sizes of all the internal angles, sorted by their size
```

I would actually see a sorting by nodes if it were a utility

GiudGiud replied:
I dont see the need for it to be a util though
do you? do you have further plans with this?

miaoyinb replied:
I am think about some general trimmer in the future.

GiudGiud commented on modules/reactor/include/meshgenerators/HexagonMeshTrimmer.h:
```diff
@@ -0,0 +1,162 @@
+   * @param trimming_section_boundary_id ID of the new external boundary formed due to
+   * trimming
+   * @param external_boundary_id ID of the external boundary of the input mesh
+   * @param side_to_move which side of the mesh needs to be removed
```
is it the side of the a,b,c normal

miaoyinb replied:
ax+by+c>0 or <0 to be removed. Will revise this

GiudGiud commented on modules/reactor/include/meshgenerators/HexagonMeshTrimmer.h:
```diff
@@ -0,0 +1,162 @@
+                            const Real dis_tol = 1.0E-6);
+
+  /**
+   * Calculates the intersect Point of two given straight lines
```
```suggestion
   * Calculates the intersection Point of two given straight lines
```

GiudGiud reviewed (changes requested):
submitting before I loose internet, continuing to work on this

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+                Mesh/hex_1/ring_block_ids="15 15"
+                Mesh/hex_1/quad_center_elements=true
+                --mesh-only "patterned_pattern_trimming_tri.e"'
+    requirement ='The system shall be able to perform center trimming on a core mesh generated by double hexagonal patterning with pin centers consiting of triangular elements.'
```
```suggestion
    requirement ='The system shall be able to perform center trimming on a core mesh generated by double hexagonal patterning with pin centers consisting of triangular elements.'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+    input = 'assembly_trim.i'
+    cli_args = 'Mesh/trim/center_trimming_section_boundary="center_section"'
+    expect_err = 'this input parameter is not used if center trimming is not performed.'
+    requirement = "The system shall throw an error if the input mesh is not center trimmed but the center trimming boundary is provided."
```
```suggestion
    requirement = "The system shall throw an error if the input mesh is not to be trimmed at the center but the center trimming boundary is provided anyway."
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+    cli_args = 'Mesh/trim/center_trim_starting_index=0
+                Mesh/trim/center_trim_ending_index=10'
+    expect_err = 'the remaining mesh after center trimming defined by this parameter and center_trim_ending_index must be equal or smaller than half of the input mesh.'
+    requirement = "The system shall throw an error if the center trimming yields a mesh that is over half of the input mesh."
```
```suggestion
    requirement = "The system shall throw an error if the center trimming would yield a mesh that is over half of the input mesh, because more than half of the angular sectors are requested."
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+    expect_err = 'this parameter must be provided along with center_trim_ending_index.'
+    requirement = "The system shall throw an error if the center trimming index parameters are not given in a pair."
+  []
+  [err_peripheral_trimmible]
```
```suggestion
  [err_peripheral_trimmable]
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+                Mesh/inactive="gmg"
+                Mesh/trim/trim_peripheral_region="1 1 1 1 1 1"'
+    expect_err = 'The input mesh does not have a trimmable peripheral region.'
+    requirement = "The system shall throw an error if the input mesh cannot be peripheral trimmed."
```
```suggestion
    requirement = "The system shall throw an error if the input mesh cannot be trimmed at the periphery."
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+    input = 'assembly_trim.i'
+    cli_args = 'Mesh/trim/peripheral_trimming_section_boundary="peripheral_section"'
+    expect_err = 'this input parameter is not used if peripheral trimming is not performed.'
+    requirement = "The system shall throw an error if the input mesh is not peripheral trimmed but the peripheral trimming boundary is provided."
```
```suggestion
    requirement = "The system shall throw an error if the input mesh is not to be trimmed at the periphery but the peripheral trimming boundary is provided."
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+    input = 'err_trimmability.i'
+    cli_args = 'Mesh/inactive="hex_1 peripheral_ring"'
+    expect_err = 'The input mesh is not compatible with HexagonMeshTrimmer.'
+    requirement = "The system shall throw an error if the input mesh is not compatible."
```
```suggestion
    requirement = "The system shall throw an error if the input mesh has not been marked compatible with hexagonal mesh trimming."
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+                Mesh/trim/center_trimming_section_boundary="center_section"
+                Mesh/inactive="peripheral_ring pattern2"
+                --mesh-only "mod_center_trimming.e"'
+    requirement ='The system shall be able to trim a fraction of an assembly mesh with its peripheral regions modified through its center.'
```
```suggestion
    requirement ='The system shall be able to trim a fraction of an hexagonal assembly mesh with its peripheral regions modified through its center.'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+                Mesh/trim/center_trimming_section_boundary="center_section"
+                Mesh/inactive="peripheral_ring pattern_mod pattern2"
+                --mesh-only "center_trimming.e"'
+    requirement ='The system shall be able to trim a fraction of an assembly mesh through its center.'
```
```suggestion
    requirement ='The system shall be able to trim a fraction of an hexagonal assembly mesh through its center.'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+                Mesh/trim/peripheral_trimming_section_boundary="peripheral_section"
+                Mesh/inactive="peripheral_ring pattern_mod pattern2"
+                --mesh-only "peripheral_trimming_partial.e"'
+    requirement ='The system shall be able to trim a fraction of the six peripheral regions off an assembly mesh.'
```
```suggestion
    requirement ='The system shall be able to trim a fraction of the six peripheral regions off an hexagonal assembly mesh.'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+                Mesh/trim/peripheral_trimming_section_boundary="peripheral_section"
+                Mesh/inactive="peripheral_ring pattern_mod pattern2"
+                --mesh-only "peripheral_trimming_full.e"'
+    requirement ='The system shall be able to trim all of the six peripheral regions off an assembly mesh.'
```
```suggestion
    requirement ='The system shall be able to trim all of the six peripheral regions off an hexagonal assembly mesh.'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+                Mesh/trim/peripheral_trimming_section_boundary="peripheral_section"
+                Mesh/inactive="peripheral_ring pattern_mod pattern2"
+                --mesh-only "double_trimming.e"'
+    requirement ='The system shall be able to perform both peripheral and center trimming simultaneously.'
```
```suggestion
    requirement ='The system shall be able to perform both peripheral and center trimming simultaneously on an hexagonal assembly.'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+                Mesh/trim/center_trimming_section_boundary="center_section"
+                Mesh/inactive="peripheral_ring pattern_mod"
+                --mesh-only "patterned_pattern_trimming.e"'
+    requirement ='The system shall be able to perform center trimming on a core mesh generated by double hexagonal patterning.'
```
```suggestion
    requirement ='The system shall be able to perform center trimming on an hexagonal core mesh generated by double hexagonal patterning.'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+                Mesh/trim/center_trimming_section_boundary="center_section"
+                Mesh/inactive="peripheral_ring"
+                --mesh-only "mod_patterned_pattern_trimming.e"'
+    requirement ='The system shall be able to perform center trimming on a core mesh generated by double hexagonal patterning with the peripheral regions of assembly meshes modified.'
```
```suggestion
    requirement ='The system shall be able to perform center trimming on an hexagonal core mesh generated by double hexagonal patterning with the peripheral regions of assembly meshes modified.'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/hexagon_mesh_trimmer/tests:
```diff
@@ -0,0 +1,222 @@
+    exodiff = 'patterned_trimmed_in.e'
+    exodiff_opts = '-pedantic'
+    cli_args = '--mesh-only'
+    requirement ='The system shall be able to stitch the assembly meshes after peripheral trimming together.'
```
```suggestion
    requirement ='The system shall be able to stitch together hexagonal assembly meshes after peripheral trimming.'
```

GiudGiud reviewed (changes requested):
same deal as last post

GiudGiud commented on modules/reactor/src/meshgenerators/PatternedHexMeshGenerator.C:
```diff
@@ -154,10 +155,14 @@ PatternedHexMeshGenerator::PatternedHexMeshGenerator(const InputParameters & par
         "control_drums_azimuthal_meta", std::vector<std::vector<Real>>())),
     _position_file_name(declareMeshProperty<std::string>("position_file_name",
                                                          getParam<std::string>("position_file"))),
+    _hexagon_peripheral_trimmability(
+        declareMeshProperty<bool>("hexagon_peripheral_trimmability", !_generate_core_metadata)),
```
so the issue with declaring these trimmabilities boolean is that if someone gives you an exodus mesh to trim you cant really do it, it needs to be fresh from the reactor module or loaded from checkpoint (meshProperty is restartable iirc)

maybe there should be an option to override these booleans?

GiudGiud replied:
we'd need to add the meta data to the loaded mesh right? is there a generator for that?

miaoyinb replied:
Currently, it only works for some MGs in Reactor modules for now.

I do have some plan in mind to have a MG to assign new metadata (especially for FileMG). Hopefully it can be done in FY23.

GiudGiud replied:
ok that works. Assigning metadata will be nice for working with the FMG indeed

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,574 @@
+            ? getParam<BoundaryName>("peripheral_trimming_section_boundary")
+            : BoundaryName()),
+    _trimming_start_sector(isParamValid("center_trim_starting_index")
+                               ? (getParam<short>("center_trim_starting_index") + 9) % 12
```
+ 9?
so the convention in the code is different from the convention in the documentation?

miaoyinb replied:
It is the same. 9 is actually -3 + 12, to solve the same issue that I want -3 % 12 = 9.

GiudGiud replied:
the % operator being so unintuitive is an issue imo. anyone knows if we have a positive modulo somewhere @lindsayad @roystgnr  ?

roystgnr replied:
> the % operator being so unintuitive

Wow, I couldn't agree more.  I always use % to get an unsigned result (content-wise, if not type-wise), so I occasionally do the "add an upper bound multiple of n before taking the modulus" trick to cope with the fact that C/C++ doesn't seem to do Euclidean division properly ... but I'd assumed that the C/C++ behavior must be some standard-among-programmers weirdness ... until I started diving into the rathole of standards and eventually found [this table of modulo operations](https://en.wikipedia.org/wiki/Modulo_operation#In_programming_languages), a total disaster.  Among popular languages only Maple and newer versions of Scheme seem to have first-class support for Euclidean division.  There's an IEEE standard FP remainder operation, but it seems to be the least popular convention in use.  C/C++ `%` on negative numbers was *implementation-defined* until C99 and C++11.  Negative operands on `%` work like C/C++ in many languages, but shader languages leave it undefined, and Lua, Perl, Python, and Ruby all use a different standard.

> anyone knows if we have a positive modulo somewhere

Not in libMesh or dependencies, at least.  I can see why not, too; skimming through `git grep` results shows a whole lot of unsigned modulo operations but I couldn't find any instances where we had to use trickery first.

miaoyinb replied:
It might be negative as there is some rotation in the algorithm.

GiudGiud replied:
yes we may have end_index < start_index to get part of the domain. 

let's add a math util ?

roystgnr replied:
Sounds like the only self-documenting option to me.  Signature `std::size_t MathUtils::euclidean_mod(T1 dividend, T2 divisor)`?  Fat return type, but shaving it down is probably non-trivial and probably irrelevant for an inlineable function.

`euclidean_mod` is the best name I can think of, but I just noticed that [this reference code for std::div](https://en.cppreference.com/w/cpp/numeric/math/div) calls a function `euclidean_division_string`, shows the example output with negative dividends returning negative remainders, then links right to a wiki page with $0≤r<|b|$ in the first section and negative dividends returning positive remainders in the fourth ... so perhaps that "Euclidean" terminology still isn't as clear as it should be.

GiudGiud reviewed (changes requested):
ok done for today, entering YST soon

miaoyinb commented on modules/reactor/include/meshgenerators/HexagonMeshTrimmer.h:
```diff
@@ -0,0 +1,162 @@
+  std::vector<unsigned short> _trim_peripheral_region;
+  /// Name of the section formed by peripheral trimming
+  const BoundaryName _peripheral_trimming_section_boundary;
+  /// Index of the sector to start trimming from (counter-clockwise direction)
+  const short _trimming_start_sector;
+  /// Index of the sector to start trimming to (counter-clockwise direction)
+  const short _trimming_end_sector;
```
Some rotation needs a negative index.

miaoyinb replied:
This fixed a bug I mentioned before.

miaoyinb commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,574 @@
+  if (_trimming_start_sector == 12 && _trimming_end_sector == 12)
+    _center_trim_sector_number = 12;
+  else
+    _center_trim_sector_number = (_trimming_end_sector - _trimming_start_sector + 12) % 12;
```
-3 % 12 = -3, I want it to be 9.

miaoyinb committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

miaoyinb committed:
doc fix ref. #21853

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,574 @@
+      {
+        if ((*elem_it)->neighbor_ptr(i) != nullptr)
+        {
+          boundary_info.add_side((*elem_it)->neighbor_ptr(i),
```
this isnt covered. Any meshes in the reactor module which would trigger that?

miaoyinb replied:
Typical reactor module meshes might not trigger this. 
lineRemover() has been designed for more general trimming purposes, which would lead to zero volume elements. I would like to keep these lines for future expansion.

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,574 @@
+    if (side_id_0_in && side_id_1_in)
+      // The side needs to be removed from the sideset
+      // The other node will be handled by other element's side
+      boundary_info.remove_side(mesh.elem_ptr(std::get<0>(slc_bdry_side_list[i])),
```
this isnt covered. Any meshes in the reactor module which could trigger this one?
You d need less than 12th symmetry right?

miaoyinb replied:
Yes, same as the previous comment.

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,574 @@
+                                     "Index of the ending center trimming position.");
+  params.addParam<BoundaryName>("center_trimming_section_boundary",
+                                "Boundary formed by center trimming.");
+  params.addParam<BoundaryName>("external_boundary", "External boundary of the input mesh.");
```
It's not used in testing.

miaoyinb replied:
This is the external boundary of the input mesh. As we want to move nodes on the external boundary along the boundary instead of in the normal direction.

miaoyinb replied:
I am adding a test for this parameter.

Also, for center trimming, the new boundary might not be the external boundary

But for the peripheral trimming, the new boundary is always the external boundary.

So I am treating them differently.

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,574 @@
+  for (const auto & bad_elem : bad_elems_rec)
+  {
+    std::vector<boundary_id_type> elem_bdry_container;
+    // elems 2 and 3 are the neighboring elements of the degenerate element corresponsing to the two
```
```suggestion
    // elems 2 and 3 are the neighboring elements of the degenerate element corresponding to the two
```

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,574 @@
+  mesh.contract();
+  mesh.prepare_for_use();
+
+  // Assign customized outer surface boundary id and name
+  if (external_boundary_id != OUTER_SIDESET_ID)
+    MooseMesh::changeBoundaryId(mesh, OUTER_SIDESET_ID, external_boundary_id, false);
```
this should not be part of the lineRemover routine. 
The first time you run this routine, you set the external_bid everywhere, then the second time it is found for the slc_bdry_list

GiudGiud commented on modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C:
```diff
@@ -0,0 +1,574 @@
+  boundary_info.build_node_list_from_side_list();
+  auto bdry_node_list = boundary_info.build_node_list();
+  // Only select the boundaries_to_conform
+  std::vector<std::tuple<dof_id_type, boundary_id_type>> slc_bdry_node_list;
+  for (unsigned int i = 0; i < bdry_node_list.size(); i++)
+    if (std::get<1>(bdry_node_list[i]) == external_boundary_id)
+      slc_bdry_node_list.push_back(bdry_node_list[i]);
```
this is not used anywhere

miaoyinb replied:
removed

GiudGiud commented on modules/reactor/doc/content/source/meshgenerators/HexagonMeshTrimmer.md:
```diff
@@ -0,0 +1,44 @@
+When trimming a mesh, some elements may be located across the trimming line and thus need to be processed to ensure a smooth trimming boundary. Generic tools such as [`PlaneDeletionGenerator`](/PlaneDeletionGenerator.md) is capable of trimming meshes but leaves a zigzag trimming boundary in the presence of across-trimming-line elements.
+
+To avoid this zig-zag boundary, `HexagonMeshTrimmer` adopts a post-trimming processing algorithm to smooth the trimming boundary. The algorithm moves the nodes of the across-trimming-line elements in the normal direction of the trimming line onto the trimming line. During this procedure, some elements may become zero volume and will be removed. More importantly, after node moving, some quadrilate ...
+
```
Add a small paragraph to explain the optional external boundary treatment, users could miss this

miaoyinb committed:
code cleanup and improvement ref #21853

miaoyinb committed:
Add and use euclidean mod ref. #21853

miaoyinb commented:
Added euclideanMod() to `MathUtils`

miaoyinb commented:
> nice. Thanks for addressing the review. I ll try to do a final pass on this. I think we're going to have coverage issues overall because the reactor meshes are too good for what this trimmer can do.
> 
> To address this, the reactor meshes should be "damaged" using the moveNodeGenerator before trimming.

Thanks, @GiudGiud. The current 95.2% coverage is a good number, right?

GiudGiud commented:
Though the number is really impressive, it s more about the features tested.

miaoyinb commented:
> nice. Thanks for addressing the review. I ll try to do a final pass on this. I think we're going to have coverage issues overall because the reactor meshes are too good for what this trimmer can do.
> 
> To address this, the reactor meshes should be "damaged" using the moveNodeGenerator before trimming.

The needed metadata would be lost after moveNodeGenerator. I need to try to see if peripheral modification of an assembly mesh could trigger this.

miaoyinb committed:
cover zero element removal ref #21853

miaoyinb commented:
> > nice. Thanks for addressing the review. I ll try to do a final pass on this. I think we're going to have coverage issues overall because the reactor meshes are too good for what this trimmer can do.
> > To address this, the reactor meshes should be "damaged" using the moveNodeGenerator before trimming.
> 
> The needed metadata would be lost after moveNodeGenerator. I need to try to see if peripheral modification of an assembly mesh could trigger this.

I managed to cover the zero volume element. The other one can only be triggered if an interface boundary (instead of external boundary) needs to be conformed, which is not the case here. The situation is for more general cases.

moosebuild commented:
Job [Framework increased sparsity](https://civet.inl.gov/job/1172599/) on 0a01125 : invalidated by @miaoyinb

miaoyinb committed:
fix uncertain mesh output due to precision ref #21853

miaoyinb commented on framework/src/utils/FillBetweenPointVectorsTools.C:
```diff
@@ -290,7 +290,7 @@ elementsCreationFromNodesVectors(ReplicatedMesh & mesh,
       Real dis1 = (*nodes[i + 1][nodes_up_it] - *nodes[i][nodes_down_it + 1]).norm();
       Real dis2 = (*nodes[i + 1][nodes_up_it + 1] - *nodes[i][nodes_down_it]).norm();
-      if (dis1 > dis2)
+      if (MooseUtils::absoluteFuzzyGreaterThan(dis1, dis2))
```
The original expression sometimes causes different meshes when using different numbers of cores due to the precision issue.

GiudGiud replied:
nice thanks for the bug fix!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1173343/) on 35dbdf8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21863/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21863/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9dcc01b396e7a63a860951dd94eba66ecd07eeb5`

GiudGiud reviewed (dismissed):
Thanks for the contribution, this is good to go!

I ll merge tomorrow when civet is back online.

GiudGiud commented:
Should have merged last night, large media got updated.
Can you please push a submodule update?

miaoyinb commented:
👍
